File: packages/guards/src/roleplay-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RoleplayDetectOptions {
  action: 'block' | 'warn';
}

const ROLEPLAY_PATTERNS: RegExp[] = [
  /\bpretend (?:you are|to be|you're)\b/gi,
  /\byou are now\s+\w+/gi,
  /\bact as (?:a|an|if you were)\b/gi,
  /\broleplay as\b/gi,
  /\bplay the role of\b/gi,
  /\bimagine you(?:'re| are) (?:a|an)\b/gi,
  /\bfrom now on,? you(?:'re| are)\b/gi,
  /\brespond as (?:a|an|if)\b/gi,
  /\bsimulate (?:being|a)\b/gi,
  /\byour new persona is\b/gi,
  /\bforget (?:that )?you(?:'re| are) an AI\b/gi,
  /\bDAN\b/g,
  /\bdo anything now\b/gi,
];

export function roleplayDetect(options: RoleplayDetectOptions): Guard {
  return {
    name: 'roleplay-detect',
    version: '0.1.0',
    description: 'Detect roleplay and persona manipulation attempts',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of ROLEPLAY_PATTERNS) {
        const re = new RegExp(p.source, p.flags);
        const m = re.exec(text);
        if (m) matched.push(m[0]);
      }
      const triggered = matched.length > 0;
      return {
        guardName: 'roleplay-detect', passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered ? `Roleplay/persona manipulation: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Input attempts to make the AI adopt a different persona' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/roleplay_detect.py
"""Detect roleplay and persona manipulation."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_PATTERNS = [
    "let's pretend", "lets pretend",
    "imagine you are", "role: ", "character:",
    "in-character", "stay in role",
    "act as if", "pretend to be",
    "you are now", "from now on you are",
]

_ACTION_RE = re.compile(r"\*[^*]+\*")


class _RoleplayDetect:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "roleplay-detect"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        reasons: list[str] = []

        matched = [p for p in _PATTERNS if p in lower]
        if matched:
            reasons.append("roleplay-keywords")

        actions = _ACTION_RE.findall(text)
        if len(actions) >= 2:
            reasons.append("asterisk-actions")

        triggered = len(reasons) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="roleplay-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Roleplay patterns detected"
                if triggered else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "reasons": reasons,
                "matched_keywords": matched,
                "action_count": len(actions),
            } if triggered else None,
        )


def roleplay_detect(
    *, action: str = "warn",
) -> _RoleplayDetect:
    return _RoleplayDetect(action=action)
